flink-filesystems/flink-gs-fs-hadoop/src/main/java/org/apache/flink/fs/gs/writer/GSRecoverableFsDataOutputStream.java

```diff
@@ -0,0 +1,177 @@
+            GSRecoverableWriter writer,
+            GSRecoverableWriterState state) {
+        this.storage = Preconditions.checkNotNull(storage);
+        this.options = Preconditions.checkNotNull(options);
+        this.writer = Preconditions.checkNotNull(writer);
+        this.state = Preconditions.checkNotNull(state);
+    }
+
+    @Override
+    public long getPos() throws IOException {
+        return state.bytesWritten;
+    }
+
+    @Override
+    public void write(int byteValue) throws IOException {
+        byte[] bytes = new byte[] {(byte) byteValue};
+        write(bytes);
+    }
+
+    @Override
+    public void write(@Nonnull byte[] content) throws IOException {
+        Preconditions.checkNotNull(content);
+
+        write(content, 0, content.length);
+    }
+
+    @Override
+    public void write(@Nonnull byte[] content, int start, int length) throws IOException {
+        Preconditions.checkNotNull(content);
+        Preconditions.checkArgument(start >= 0);
+        Preconditions.checkArgument(length >= 0);
+
+        // if the data stream is already closed, throw an exception
+        if (state.closed) {
+            throw new IOException("Illegal attempt to write to closed output stream");
+        }
+
+        // if necessary, create a write channel
+        if (currentWriteChannel == null) {
+            currentWriteChannel = createWriteChannel();
+        }
+
+        // write to the stream. the docs say that, in some circumstances, though an attempt will be
+        // made to write all of the requested bytes, there are some cases where only some bytes will
+        // be written. it's not clear whether this could ever happen with a Google storage
+        // WriteChannel; in any case, recoverable writers don't support partial writes, so if this
+        // ever happens, we must fail the write.:
+        // https://docs.oracle.com/javase/7/docs/api/java/nio/channels/WritableByteChannel.html#write(java.nio.ByteBuffer)
+        int bytesWritten = currentWriteChannel.write(content, start, length);
+        if (bytesWritten != length) {
+            throw new IOException(
+                    String.format(
+                            "WriteChannel.write wrote %d of %d requested bytes, failing.",
+                            bytesWritten, length));
+        }
+
+        // update count of total bytes written
+        state.bytesWritten += length;
+    }
+
+    @Override
+    public void flush() throws IOException {
+        // not supported for GS, flushing frequency is controlled by the chunk size setting
+        // https://googleapis.dev/java/google-cloud-clients/0.90.0-alpha/com/google/cloud/WriteChannel.html#setChunkSize-int-
+    }
+
+    @Override
+    public void sync() throws IOException {
+        // not supported for GS, flushing frequency is controlled by the chunk size setting
+        // https://googleapis.dev/java/google-cloud-clients/0.90.0-alpha/com/google/cloud/WriteChannel.html#setChunkSize-int-
+    }
```

Comment:
Agreed, these methods don't seem to be called -- I set breakpoints and watched it run, and I never saw them get hit.

Now that I look at this again, I realize that the current noop implementation is really a holdover from the original implementation we considered, where a single ```WriteChannel``` would be used for the entire upload. In that mode, there's really no way to flush or sync. 

However, since we've changed it so that we use multiple write channels, I think I could implement ```flush``` and ```sync``` by closing the active write channel (if it exists). So, basically just implement each of those methods by calling the private method ```closeChannelIfExists```.

Shall I just do that? That way, if the core recoverable writer code ever does start calling one of those methods, we'll have an appropriate implementation in place.

Reply:
I'd be ok with either way. It's up to you.
- It would surely be nice if we can have these methods properly implemented, if that does not add too much workload for you.
- I'd also be fine with leaving them unsupported for and implement them later when they are needed. In that case, we need to well document that the methods are not implemented only because they are not used, rather than any limitations from GCS.

Reply:
They'd be trivial to implement, so I'll just go ahead and do that.

Reply:
Done in [34ad465](https://github.com/apache/flink/pull/15599/commits/34ad465ca419b0afdb399e53deca8b27027f7191).